zds/utils/templatetags/topbar.py

```diff
@@ -55,39 +55,62 @@ def top_categories(user):
 
 @register.filter('top_categories_content')
 def top_categories_content(_type):
-        """Get all the categories and their related subcategories associated with existing contents.
-        The result is sorted by alphabetic order.
-
-        :param _type: type of the content
-        :type _type: str
-        :return: a dictionary, with the title being the name of the category and the content a list of subcategories,
-        Each of these are stored in a tuple of the form ``title, slug``.
-        :rtype: OrderedDict
-        """
-
+    """Get all the categories and their related subcategories associated with existing contents.
+    The result is sorted by alphabetic order.
+
+    :param _type: type of the content
+    :type _type: str
+    :return: a dictionary, with the title being the name of the category and the content a list of subcategories,
+    Each of these are stored in a tuple of the form ``title, slug``.
+    :rtype: OrderedDict
+    """
+    # get subcategories from PublishedContent
+    if _type:
         subcategories_contents = PublishedContent.objects\
             .filter(content_type=_type)\
             .values('content__subcategory').all()
+    else:
+        # used in page with all content types
+        subcategories_contents = PublishedContent.objects\
+            .values('content__subcategory').all()
 
-        categories_from_subcategories = CategorySubCategory.objects\
-            .filter(is_main=True)\
-            .filter(subcategory__in=subcategories_contents)\
-            .order_by('category__position', 'subcategory__title')\
-            .select_related('subcategory', 'category')\
-            .values('category__title', 'subcategory__title', 'subcategory__slug')\
-            .all()
+    # get parent categories of subcategories from PublishedContent
+    categories_from_subcategories = CategorySubCategory.objects\
+        .filter(is_main=True)\
+        .filter(subcategory__in=subcategories_contents)\
+        .order_by('category__position', 'subcategory__title')\
+        .select_related('subcategory', 'category')\
+        .values('category__title', 'subcategory__title', 'subcategory__slug')\
+        .all()
+
+    # store all categories in a dict with only title and slug
+    cats = OrderedDict()
+    for csc in categories_from_subcategories:
+        key = csc['category__title']
+
+        if key in cats:
+            cats[key].append((csc['subcategory__title'], csc['subcategory__slug']))
+        else:
+            cats[key] = [(csc['subcategory__title'], csc['subcategory__slug'])]
 
-        categories = OrderedDict()
+    # get all tags from contents
+    tmp_tags = (PublishedContent.objects
```

Comment:
Car l'objet publishedcontent c'est en quelque sorte l'historique de 
publication d'un contenu du coup un même contenu peut avoir plusieurs 
publishedcontent. must_redirect=True c'est comme sélectionner la 
dernière publication d'un contenu.

Le 22/03/2016 19:26, Laville Augustin a écrit :
>
> In zds/utils/templatetags/topbar.py 
> <https://github.com/zestedesavoir/zds-site/pull/3261#discussion_r57042185>:
>
> >
> > -        categories = OrderedDict()
> > +    # get all tags from contents
> > +    tmp_tags = (PublishedContent.objects
>
> pourquoi ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/zestedesavoir/zds-site/pull/3261/files/753b2255672250bba3b649f7e2644c3579d54fdd#r57042185>
>